### ISSUE TITLE
Warn users that we collect analytics on first start

### ISSUE THREAD

**Issue by jrddunbr** [eln2/eln2#187]
When the configuration file is generated for the first time _do not_ send an analytics packet and warn the user with a chat message that if they do not set it to false in the config that we will log basic information about their game's performance starting on the next run of the mod.

---

Currently we log 2 things that could be considered user-identifiers:
* uuid - a randomly generated UUID that is stored in their configuration for analytics
* IP address - the IP address that the server receives traffic from

These identifiers allow us to remove your data in the event of a GDPR (or if you just want to) request. Simply shoot a developer a message with your IP address(es) and/or UUID(s) from your config and we will delete _everything_ we have on your user.

The uuid field is also used as a dedup flag, and the IP address may be used for spam reduction purposes.

We also log the following currently (but may include more in the future - the source code is an authoritative reference):
* mod_id - what mod you're playing from Age Series (probably should also add version number?)
* game_lang - the language configured in the game
* os_lang - the language configured on the computer you are using (to find if we should be supporting more languages than people set in their Minecraft configuration)
* avg_sim_performance - average tick time of the MNA I guess? This is mostly a test field at this point but could combine all sims performance? idk.

Data is processed and stored in the AWS us-east-1 region, and besides the data collected itself, the various logs are configured to not be stored longer than a day on the components of the serverless architecture that we are using.

**Comment by jrddunbr** [comment]
@empireu Since you're the most critical of this, do you have any other suggestions to protect user's privacy short of disabling analytics by default?

**Comment by MrBigss** [comment]
At minimum, the check for the "first packet" should be tracked "per user" and "per machine".  
Since a lot of pack makers run the pack, change the configuration, and then distribute the pack, it stand to reason that a pack maker would set that to "it's ok", and then the user would be automatically opted in, breaching their privacy.  
On the reverse, a pack maker could also opt out users that would be ok with this.  
  
Personally, and this is where I think empireu is not the most critical of this, I think people should always get explicit consent from others, for anything, and not bury it in a "if I do this to you, and you don't object the first time, then I assume it's ok to do it to you again".
  
Making things opt-out instead of opt-in, not making it easy for people to opt out of things, tricking into things, or coercing into things and the like gives me too much of a rape vibe kinda feeling and doesn't feel ethical.  
  
The option I would like to see would be an opt-in check box in the options or similar, easy to find and check, with periodic reminders that analytic would greatly help the developers, which can also be turned off.

**Comment by jrddunbr** [comment]
This is a fair point.

I don't want to put unique identifiers into the config to re-prompt users if it detects something like a different hostname or anything like that; I'm kinda conflicted but I know that if this isn't somehow opt-out that we will never get any useful information.

I definitely like the idea of doing a configuration option in the GUI screen.

**Comment by MrBigss** [comment]
The config could more simply use the minecraft username. Like having a hash of which users have agreed in the mod's config.  
(One way, so you can't get the pack maker's uuid either if they distribute their config with their pack.)    
This would satisfy the "per machine". I didn't mean it as in "do the Microsoft Windows activation check scheme", simply don't go the other way either, and use  a database of minecraft users who have agreed.

And I do understand that it is hard to get people on board. So the reason why people do not want to be on board must be addressed. And the benefit for the mod makers also exposed.  
The exact information collected, the reasons for collecting that information, and the intent behind the collection must all be shown upfront. We have the luxury of also being able to point to the code, and show exactly where this happens, what happens, and what is sent.  
  
I think getting people on board a Patreon is exceptionally harder, but still done, so, explaining how metrics do help the mod, without a significant cost to the player, might be enough to entice enough players to participate.  
  
Obviously, my practical side also says "opt-out".  
But my ethical side always said all metrics should be "opt-in".  
  
Are there any cosmetic incentive that could be given to people opting-in?  
Could the metrics themselves be used somewhere in the mod, adding content to those who opt-in?  
Maybe people who opt-in also have the option of opting into having their name on a "metrics supporter" mention or something?  
  
I think the reticence of people towards metrics in itself is a symptom of a problem (of trust), and that this could really benefit from an innovative approach for the whole domain of analytic.  
  
How do you get people to want to give you metrics, instead of trying to make them "agree by omission", and not thinking about it?  
How do you make people trust what you are going to do with the metrics?  
How do you get people to see it's for their benefit to send the metrics? (It's a benefit for the users if it benefit the mod maker, since making the mod better benefits the users.)  
  
An example would be to expose how having analytics on which systems in the mods are being used help balance the mod itself.  
Like, let's say that a certain power generation item is being skipped in "tier 2" and "tier 1" get spammed instead, it can be shown through analytics that "tier 2 power" is not "good enough" for people to want to use it, and thus balance it out until it's adoption is what is expected.  

I'm sure there's a lot of innovation waiting in analytics, we should tap into that potential.

**Comment by jrddunbr** [comment]
> The config could more simply use the minecraft username. Like having a hash of which users have agreed in the mod's config.
(One way, so you can't get the pack maker's uuid either if they distribute their config with their pack.)
This would satisfy the "per machine". I didn't mean it as in "do the Microsoft Windows activation check scheme", simply don't go the other way either, and use a database of minecraft users who have agreed.

This sounds overly excessive and I don't want to track a list of people who have agreed. I was thinking of doing something like taking the hash of their primary nic's MAC address, but to be honest these kinds of things can be reversed pretty quickly because the search space is too small, even with salts. Usernames are much easier to try and hash to reverse them because it's trivial to find usernames of common players and hash them; it's not like passwords where it's not generally public and with a more complicated search space (since passwords often don't have words or other things in them).

TL;DR Security theater shows that usernames and MAC addresses are likely terrible identifiers and are more likely to be a direct identifier than just using a randomly generated uuid.

> I think getting people on board a Patreon is exceptionally harder, but still done, so, explaining how metrics do help the mod, without a significant cost to the player, might be enough to entice enough players to participate.

We're not looking for _any_ financial incentive, and I (nor anyone else on this project) would want to be bound to doing this in spare time for a financial incentive. There's also complicated tax implications and other reasons to which I won't discuss but funding the mod has been a question for years and my general sentiment is that I don't need nor want your money; I do this for fun and if it wasn't fun I wouldn't be doing it in the first place; Patreon might just make it pay a little.

As a side node.. Eln2 developer pizza? :P

We joked about if we got a quarter every time someone asked about 1.12 and discovered that we could buy the developers a pizza with the quarters if we had. I digress...

> Obviously, my practical side also says "opt-out".
But my ethical side always said all metrics should be "opt-in".

I agree, I usually opt out as soon as I get something myself.. but we're not really selling data (nevermind personal data) or anything like that.

> Are there any cosmetic incentive that could be given to people opting-in?
Could the metrics themselves be used somewhere in the mod, adding content to those who opt-in?

Oh certainly, we could have some global performance graph. My guess is that any hardcore user would also probably set up Graphana/Elasitcsearch/OpenSearch in order to be able to have better metrics on their performance anyways.

> Maybe people who opt-in also have the option of opting into having their name on a "metrics supporter" mention or something?

I would certainly not want to gate things behind supporting metrics, I hate loot boxes and dislike the cosmetics craze personally.

> I think the reticence of people towards metrics in itself is a symptom of a problem (of trust), 

1. People have no reason to trust me. And I don't expect them to - if they don't want to send their data, they don't have to send it.
2. The infrastructure used to collect this analytics is currently using AWS with customer encryption via AWS-provided KMS keys.
   * If you want to read more about how they do that, [read up on their data privacy page](https://aws.amazon.com/compliance/data-privacy-faq/), and the TLS certificates that are used on the service are going to either be AWS keys or LetsEncrypt keys.
3. I currently do not intend to set this up in multiple geographic regions, but in the future if the need arises the infrastructure can be scaled to multiple geographic regions and comply with data privacy laws that probably don't apply to us but are still nice to have.
    * If this is a concern, we should probably work this out before launch. I am not a lawyer and I don't have one for this project.
    * The infrastructure is pretty flexible and is comprised of a single DNS record, an API Gateway, a Lambda, and a DynamoDB table based in AWS's `us-east-1` region in the USA.
    * This could be scaled to multiple geographical regions with just the addition of country-based DNS records (perhaps based off of the language you select or a configuration option, eg. the `en-us` language selection would send to a USA based server), another API gateway for that hostname, and probably just make the same lambda in multiple regions and same with the table.
    * Worth noting that I'm still going to need to pull the data into the USA because that's where I actually process them in a spreadsheet program on my computer.
    * The infrastructure does log your IP address for the sake of spam management but if you're upset with us knowing your IP address you should probably consider using a VPN instead because your ISP is going to do _far_ worse things with that data than we will ever be able to do.
4. If you ask for your data to be deleted, I will delete it for you. I'm not going to be a nuisance about it as long as you provide me with an identifier to know _which_ data to delete.

> and that this could really benefit from an innovative approach for the whole domain of analytic.

I don't have any innovative ideas on the subject but if you have a better idea of getting these metrics from people we don't normally interact with, I'd be all ears. The issue is that we get a lot of people who don't bother to reach out to us and otherwise don't know to reach out to us who could still benefit from these analytics being used as a metric to show what improvements would be best; asking the local population never hurts but the point of this is to:

1. Get an idea of how much of a userbase we have to understand how much time I should invest
2. How much of it we might be engaging via support threads
3. What issues people might be encountering that are not a part of the userbase that we regularly see in the Discord forums.

> How do you get people to want to give you metrics, instead of trying to make them "agree by omission", and not thinking about it?

Ask/notify nicely and respect their privacy when they do reach out.

> How do you make people trust what you are going to do with the metrics?

1. We don't do anything bad with metrics (this is a really bad generic answer, I get it)
2. I generally share the metrics with the team/anyone in the Discord at a high level, because they are anonymized.
3. The only person with direct access to the raw metrics and the metrics system is myself, and this is unlikely to change. (for Eln2 - my fork of Eln1 also sent metrics upstream to the original analytics in addition to myself - we don't do that in the new version)
4. I'm totally open to having discussions about the topic and open to share what we do with it and be as transparent as possible.

> How do you get people to see it's for their benefit to send the metrics? (It's a benefit for the users if it benefit the mod maker, since making the mod better benefits the users.)

Thanks for helping the team! The current largest benefit I foresee is the ability to identify what languages we should support localization for and which ones to prioritize trying to find people for. We could call this out in a message someplace when the user logs in.

**Comment by jrddunbr** [comment]
I did some research on the GDPR. A lot of stuff doesn't apply to us because:
1. We are less than 250 people and doesn't primarily process user data, which means [we have no requirements for record keeping](https://gdpr.eu/article-30-records-of-processing-activities/).
2. We are not a formal organization, we are a collection of people who are not collected under any recognized organized legal structure (eg. we don't have a LLC, we are not a 501(c)3, we are not a company), it's just me and my friends..
2. If we don't collect IP and UUID's, [our data does not apply to the GDPR because it's completely anonymized](https://gdpr.eu/Recital-26-Not-applicable-to-anonymous-data/), and we also [don't need to track data about users enough to provide their data](https://gdpr.eu/article-11-what-personal-data-can-a-controller-process-without-identification/) if we don't collect identifying information to even find it. Honestly, we don't really have much data to share with the individual user that they couldn't already collect themselves, so it's a moot point to be able to provide a single user's data.

I'm feeling inclined to just not track the user's IP address/random UUID, and then we don't need to actually worry about this at all.

We _might_ want to track country of origin for data sovereignty laws, but that is not PII in GDPR. We can collect that information from the IP address and/or the language localization from their computer, and then delete the IP address since that anonymizes the data we get out of the system.

Data sovereignty doesn't protect the data of individuals PII except in Russia and China, but I am guessing that we would just discard the data from those nations anyhow; there are no usable AWS regions (by me) in either country, so it's unlikely to be anything we would plan supporting.

**Comment by jrddunbr** [comment]
Whoa, got an idea. User opts into sending username and email and in return we can send a monthly email with their statistics.

Aggregated over a month we could probably provide interesting carbon footprint statistics or other curious things about the simulator over long periods of time.

I think a weekly or monthly report to the users in the Discord (via a web hook) about the statistics gathered and results such as languages used would be really cool.

**Comment by MrBigss** [comment]
>> The config could more simply use the minecraft username. Like having a hash of which users have agreed in the mod's config.
>> (One way, so you can't get the pack maker's uuid either if they distribute their config with their pack.)
>> This would satisfy the "per machine". I didn't mean it as in "do the Microsoft Windows activation check scheme", simply don't go the other way either, and use a database of minecraft users who have agreed.

>This sounds overly excessive and I don't want to track a list of people who have agreed. I was thinking of doing something like taking the hash of their primary nic's MAC address, but to be honest these kinds of things can be reversed pretty quickly because the search space is too small, even with salts. Usernames are much easier to try and hash to reverse them because it's trivial to find usernames of common players and hash them; it's not like passwords where it's not generally public and with a more complicated search space (since passwords often don't have words or other things in them).

> TL;DR Security theater shows that usernames and MAC addresses are likely terrible identifiers and are more likely to be a direct identifier than just using a randomly generated uuid.  
  
I don't know if you misunderstood, but I meant that this would stay locally in the machine.  
Once someone opts-out, their minecraft uuid is put in ELN's config file, and that user doesn't get asked anymore.  
And that's it, not, "globally on all install everywhere on any machine" or anything like that.  
The advantage would have been that if a pack makers opts out, and distributes it, then people receiving the pack wouldn't be opted out.  
  
And as for what you'll do with analytics, I think you'll have to come up with a better plan than "We'll get analytics then see what we'll do with them".  
Seems like getting analytics for analytics's sake.

**Comment by Caeleron** [comment]
Something to note here: this mod is licensed MIT. It's not out of the question that if someone is incredibly against analytics or other telemetry, they can build the mod on their own with said functionality removed. It should not be difficult to isolate telemetry to a singular location such that a build-time switch can convert the functionality into no-ops.

**Comment by jrddunbr** [comment]
It currently actually is isolated to one file, so that's trivial to remove if you wanted. It would be very simple.

**Comment by jrddunbr** [comment]
For the record, this is the code used in the lambda that actually collects and then stores the data.

https://github.com/age-series/stats-api

Feel free to make pull requests against it, and I'll update the lambda in AWS manually.

I've posted an issue to delete all of the data before we make it GDPR compliant so that we don't need to worry about GDPR compliance anymore (since anonymized data is not subject to the GDPR or data sovereignty rights). Once I update the lambda, I'll clear all the objects in the DynamoDB table it sends to.
